### ISSUE TITLE
Test smooth_MSELoss of infer_rnn_rates, and incorporate option + config for using it

### ISSUE THREAD

**Issue by abuzarmahmood** [katzlabbrandeis/blech_clust#491]
Update `infer_rnn_rates.py`
Refer to code in https://github.com/abuzarmahmood/blechRNN/blob/master/src/train.py

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **File Identification**:
   - The file identified for modification is `/home/exouser/Desktop/blech_github_bot/src/repos/katzlabbrandeis/blech_clust/utils/infer_rnn_rates.py`. This file is responsible for using an Auto-regressive Recurrent Neural Network (RNN) to infer firing rates from electrophysiological data.

2. **Issue Description**:
   - The issue involves incorporating a new loss function `smooth_MSELoss` into the `infer_rnn_rates.py` file and providing an option and configuration for using it. The reference for this function is found in the `blechRNN` repository.

3. **Plan to Address the Issue**:
   - Define the `smooth_MSELoss` function within the `infer_rnn_rates.py` file.
   - Modify the code to allow selection between `MSELoss` and `smooth_MSELoss`.
   - Update the configuration to include an option for selecting the loss function.

4. **Specific Changes**:

   1. **Define `smooth_MSELoss` Function**:
      - **File**: `/home/exouser/Desktop/blech_github_bot/src/repos/katzlabbrandeis/blech_clust/utils/infer_rnn_rates.py`
      - **Line**: After line 100
      - **Description**: Add the `smooth_MSELoss` class definition.
      - **Code snippet**:
        ```python
        class smooth_MSELoss(nn.Module):
            """
            MSE loss with temporal smoothness constraint
            """

            def __init__(self, alpha=0.05):
                super(smooth_MSELoss, self).__init__()
                self.loss1 = nn.MSELoss()
                self.alpha = alpha

            def mean_diffrence(self, x):
                """
                Calculate the mean difference between adjacent elements

                Args:
                    x: (seq_len, batch, output_size)
                """
                return torch.mean(torch.abs(x[1:] - x[:-1])) * self.alpha

            def forward(self, input, target):
                """
                Args:
                    input: (seq_len,batch, output_size)
                    target: (seq_len,batch, output_size)
                """
                loss = self.loss1(input, target) + self.mean_diffrence(input)
                return loss
        ```

   2. **Modify Code to Allow Loss Function Selection**:
      - **File**: `/home/exouser/Desktop/blech_github_bot/src/repos/katzlabbrandeis/blech_clust/utils/infer_rnn_rates.py`
      - **Line**: Around line 473
      - **Description**: Add logic to select between `MSELoss` and `smooth_MSELoss`.
      - **Code snippet**:
        ```python
        # Add an argument for selecting the loss function
        parser.add_argument('--loss_function', type=str, choices=['mse', 'smooth_mse'],
                            default='mse', help='Loss function to use (default: mse)')

        # Update the criterion based on the selected loss function
        if args.loss_function == 'smooth_mse':
            criterion = smooth_MSELoss()
        else:
            criterion = MSELoss()

        # Pass the criterion to the train_model function
        net, loss, cross_val_loss = train_model(
            net,
            train_inputs,
            train_labels,
            output_size=output_size,
            lr=0.001,
            train_steps=train_steps,
            criterion=criterion,
            test_inputs=test_inputs,
            test_labels=test_labels,
        )
        ```

   3. **Update Configuration for Loss Function Option**:
      - **File**: `/home/exouser/Desktop/blech_github_bot/src/repos/katzlabbrandeis/blech_clust/utils/infer_rnn_rates.py`
      - **Line**: Around line 142
      - **Description**: Add logic to update the configuration with the selected loss function.
      - **Code snippet**:
        ```python
        if args.loss_function:
            print(f'Using provided loss function: {args.loss_function}')
            params_dict['loss_function'] = args.loss_function
        ```

These changes will incorporate the `smooth_MSELoss` into the `infer_rnn_rates.py` file and provide an option to select it via command-line arguments or configuration.

---
*This response was automatically generated by blech_bot using model gpt-4o*

**Comment by abuzarmahmood** [comment]
include alpha setting for smooth_MSELoss as an argument with a default setting in infer_rnn_rates.py

**Comment by abuzarmahmood** [comment]
[ develop_issue ]
